private/bufpkg/bufimage/bufimagebuild/builder.go

```diff
@@ -125,9 +130,118 @@ func (b *builder) build(
 	if err != nil {
 		return nil, nil, err
 	}
+	if err := b.warnInvalidImports(ctx, image, directDeps, localWorkspace); err != nil {
+		b.logger.Error("warn_invalid_imports", zap.Error(err))
+	}
 	return image, nil, nil
 }
 
+// warnInvalidImports checks that all the target image files have valid imports statements that come
+// from the local module target files, workspace target files, or from a direct dependency, and not
+// a transitive (or unknown) one. It outputs a WARN message per invalid import statement.
+func (b *builder) warnInvalidImports(
+	ctx context.Context,
+	builtImage bufimage.Image,
+	explicitDirectModuleDeps []bufmoduleref.ModuleReference,
+	localWorkspace bufmodule.Workspace,
+) error {
+	if explicitDirectModuleDeps == nil && localWorkspace == nil {
+		return nil
+	}
+	// prepare direct deps as a map
+	directDepsIdentities := make(map[string]string, len(explicitDirectModuleDeps))
+	for _, directDep := range explicitDirectModuleDeps {
+		directDepsIdentities[directDep.IdentityString()] = "explicit"
+	}
+	// if it's a workspace build, and there's any named module in it, then it's considered a direct
+	// dependency too
+	if localWorkspace != nil {
```

Comment:
Since workspaces are only for local development, do we still want to require workspace modules to be listed as direct dependencies (even if they are resolved using local files)?

Reply:
Good question. I don't think so. Assuming I have an `import "file/from/local/workspace/module"`, then `buf build` will succeed (no warning), and when doing `buf push` it'll fail since that dependency is not set in the `buf.yaml`. I'll check in some existing workspace tests to see if I'm missing something.

Reply:
Double checked, you're absolutely right, we should warn if it's resolved from a workspace file that's not explicitly imported. Added that logic.

Reply:
@unmultimedio If I understand correctly, to prevent the WARN log, we should always add the local modules resolved via buf.work.yaml#directories as direct dependency in the buf.yaml#deps?

Reply:
Yes @onmomo that's the suggested action. If the dependencies exist in the BSR (were ever pushed and it's intended to keep them being pushed), then they should exist in the buf.yaml#deps and you should `buf mod update` regularly to keep it in sync. If your dependency is local-only, then it should be unnamed (meaning buf.yaml#name is not set), and the WARN should not print.

Lmk if that helps.